### PR TITLE
3 improve trainer

### PR DESCRIPTION
**Pull request by AndreiXYZ**
See https://github.com/MantisAI/grants_tagger_light/pull/9

### CODE REVIEW

File: data/grants_comparison/mesh_tree_letters_list.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Information Sources: L
```
Comment (nsorros) [inline data/grants_comparison/mesh_tree_letters_list.txt:1]: so what's this?

**Comment by nsorros** [comment]
Let's get the DVC pipeline working

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+# TODO refactor the two load funcs into a class
+
+
+def _tokenize(batch, tokenizer: AutoTokenizer, x_col: str):
+    return tokenizer(
+        batch[x_col],
+        padding="max_length",
+        truncation=True,
+        max_length=512,
```
Comment (nsorros) [inline grants_tagger_light/training/dataloaders/mesh_json_loader.py]: this should depend on the tokenizer or at least be a param

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    )
+
+
+def _label_encode(batch, mesh_terms_column: str, label2id: dict):
+    batch_labels = []
+    for sample_tags in batch[mesh_terms_column]:
```
Comment (nsorros) [inline grants_tagger_light/training/dataloaders/mesh_json_loader.py]: I think `batch_tags` is cleaner, you are not sampling

Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: I'm not sure I understand what you mean here

File: data/grants_comparison/mesh_tree_letters_list.txt (hunk 1 above)
Comment (AndreiXYZ) [reply data/grants_comparison/mesh_tree_letters_list.txt:1]: This is for the evaluation. It's the result of [this discussion](https://wellcome.slack.com/archives/C02JM7DC3UP/p1687169778045799) with Kirstin. 
In the MeSH terms we were provided, we also got some higher level codes included i.e. non-leaf nodes as I understand. I've stored those inside this file.

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 2 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: You mean the arguments used for tokenization?

**Comment by AndreiXYZ** [comment]
> Let's get the DVC pipeline working

Added a pipeline in the latest commit. Tested and pushed the dvc artifact to remote storage

**Comment by nsorros** [comment]
Note that this code does not seem to run in a non GPU machine

**Comment by nsorros** [comment]
Another issue is that the model is being picked up from the hub, which means the code does not work in the absence of the model in the hub. Instead the model should be picked up from the local definition.

**Comment by nsorros** [comment]
At the moment I am getting this error in a `g5.2xlarge` instance 
```
RuntimeError: `fused=True` requires all the params to be CUDA, floating point Tensor
ERROR: failed to reproduce 'pipelines/bertmesh/dvc.yaml:train': failed to run: python ../../scripts/train_bertmesh_script.py --model_key Wellcome/WellcomeBertMesh --data_path ../../data/raw/allMeSH_2021.json --output_dir ../../bertmesh_outs/pipeline_test --max_samples 10 --per_device_train_batch_size 4 --per_device_eval_batch_size 4, exited with 1
```

**Review by nsorros — CHANGES_REQUESTED** [review]
- Let's get dvc repro working
- Let's use the model defined locally not from the hub
- Remove redundant train script

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 2 above)
Comment (nsorros) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: yes

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 3 above)
Comment (nsorros) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: rename `sample_tags` to `batch_tags`

File: scripts/train_bertmesh_script.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import numpy as np
```
Comment (nsorros) [inline scripts/train_bertmesh_script.py]: we do not want another script, we want dvc to run the train file, this is redundant. remove.

Comment (AndreiXYZ) [reply scripts/train_bertmesh_script.py]: Removed

**Comment by AndreiXYZ** [comment]
> 

The code works by loading a model from a model key. This can be either a model defined locally or a local path. Both work.

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 2 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: Since we only support bertmesh for the foreseeable future I think it's safe to leave these arguments hardcoded. I'm thinking, why would you want to ever change these?

File: grants_tagger_light/training/train.py [hunk 5]
```diff
@@ -108,4 +108,21 @@ def train_bertmesh_cli(
 
 
 if __name__ == "__main__":
-    train_app()
+    from dataclasses import dataclass
+
+    @dataclass
+    class TrainFuncArgs:
```
Comment (nsorros) [inline grants_tagger_light/training/train.py:167]: why is that needed and we cannot trigger the `train_app`?

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 3 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: Here I use the word sample to refer to an example in the dataset. Since I'm iterating over the batch, naming the iteration variable `batch_tags` would not be appropriate I think

**Comment by AndreiXYZ** [comment]
> Note that this code does not seem to run in a non GPU machine

What problem are you encountering?

File: grants_tagger_light/training/train.py (hunk 5 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/train.py:167]: That could be possible but then we'd need to run this from dvc, and I don't know if `typer` works the same when invoking directly as a script.

Comment (AndreiXYZ) [reply grants_tagger_light/training/train.py:167]: Just tested and it won't detect the `ctx` with the extra args. I would keep it like this for now.

**Comment by AndreiXYZ** [comment]
> Another issue is that the model is being picked up from the hub, which means the code does not work in the absence of the model in the hub. Instead the model should be picked up from the local definition.

The code works by loading a model from a model key. This can be either a model defined locally or a local path. Both should work.

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 3 above)
Comment (nsorros) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: `example_tags` could be cleaner but up to you

**Comment by nsorros** [comment]
> > Another issue is that the model is being picked up from the hub, which means the code does not work in the absence of the model in the hub. Instead the model should be picked up from the local definition.
> 
> The code works by loading a model from a model key. This can be either a model defined locally or a local path. Both should work.

Let's switch the pipeline to refer to something local but when I say local I mean to use the definition from the local file `grants_tagger_light/models/bert_mesh/model.py`. No model should be require to be present or in the hub to train. We just need code.

**Comment by AndreiXYZ** [comment]
> At the moment I am getting this error in a `g5.2xlarge` instance
> 
> ```
> RuntimeError: `fused=True` requires all the params to be CUDA, floating point Tensor
> ERROR: failed to reproduce 'pipelines/bertmesh/dvc.yaml:train': failed to run: python ../../scripts/train_bertmesh_script.py --model_key Wellcome/WellcomeBertMesh --data_path ../../data/raw/allMeSH_2021.json --output_dir ../../bertmesh_outs/pipeline_test --max_samples 10 --per_device_train_batch_size 4 --per_device_eval_batch_size 4, exited with 1
> ```

Should be fixed in the latest commit

**Comment by AndreiXYZ** [comment]
> > > Another issue is that the model is being picked up from the hub, which means the code does not work in the absence of the model in the hub. Instead the model should be picked up from the local definition.
> > 
> > 
> > The code works by loading a model from a model key. This can be either a model defined locally or a local path. Both should work.
> 
> Let's switch the pipeline to refer to something local but when I say local I mean to use the definition from the local file `grants_tagger_light/models/bert_mesh/model.py`. No model should be require to be present or in the hub to train. We just need code.

That's actually not trivial to implement in huggingface. Instantiating a model from scratch requires some config files i.e. we would have to copy all of these into our repo: 
https://huggingface.co/Wellcome/WellcomeBertMesh/tree/main

Unless I am missing something, of course

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 3 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: Fair. Renamed to `example_tags`

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 2 above)
Comment (nsorros) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: Try a different base model

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    # Split into train and test
+    dset = dset.train_test_split(test_size=test_size)
+
+    return dset["train"], dset["test"], label2id
+
+
+if __name__ == "__main__":
+    from transformers import AutoModel
+
+    model = AutoModel.from_pretrained(
```
Comment (nsorros) [inline grants_tagger_light/training/dataloaders/mesh_json_loader.py]: is that used?

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        "Wellcome/WellcomeBertMesh", trust_remote_code=True
+    )
+    tokenizer = AutoTokenizer.from_pretrained("Wellcome/WellcomeBertMesh")
```
Comment (nsorros) [inline grants_tagger_light/training/dataloaders/mesh_json_loader.py]: also instantiate from local instead of the hub

File: grants_tagger_light/training/train.py [hunk 8]
```diff
@@ -129,27 +112,66 @@ def sklearn_metrics(prediction: EvalPrediction):
 
     metrics = trainer.evaluate(eval_dataset=val_dset)
 
-    print(metrics)
+    logger.info(pformat(metrics))
```
Comment (nsorros) [inline grants_tagger_light/training/train.py:118]: do we write the metrics anywhere other than the logs? if not lets do

Comment (nsorros) [reply grants_tagger_light/training/train.py:118]: one option is to include them in the config, maybe there are there already?

Comment (AndreiXYZ) [reply grants_tagger_light/training/train.py:118]: They are logged to wandb

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 7 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: Removed as it was not used

File: grants_tagger_light/training/dataloaders/mesh_json_loader.py (hunk 6 above)
Comment (AndreiXYZ) [reply grants_tagger_light/training/dataloaders/mesh_json_loader.py]: No

File: grants_tagger_light/training/train.py (hunk 8 above)
Comment (nsorros) [reply grants_tagger_light/training/train.py:118]: we need them somewhere in the repo, we dont use wandb to find results. so they either need to be recorded in dvc or in the config.
